prow/statusreconciler/status.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package statusreconciler
+
+import (
+	"context"
+	"io/ioutil"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/pkg/io"
+	"k8s.io/test-infra/prow/config"
+	"sigs.k8s.io/yaml"
+)
+
+type storedState struct {
+	config.Config
+}
+
+type statusClient interface {
+	Load() (chan config.Delta, error)
+	Save() error
+}
+
+type statusController struct {
+	logger        *logrus.Entry
+	opener        io.Opener
+	statusURI     string
+	configPath    string
+	jobConfigPath string
+
+	storedState
+	config.Agent
+}
+
+func (s *statusController) Load() (chan config.Delta, error) {
+	s.Agent = config.Agent{}
+	state, err := s.loadState()
+	if err == nil {
+		s.Agent.Set(&state.Config)
+	}
+	changes := make(chan config.Delta)
+	s.Agent.Subscribe(changes)
+
+	if err := s.Agent.Start(s.configPath, s.jobConfigPath); err != nil {
+		s.logger.WithError(err).Fatal("Error starting config agent.")
+		return nil, err
+	}
+	return changes, nil
+}
+
+func (s *statusController) Save() error {
+	if s.statusURI == "" {
+		return nil
+	}
+	entry := s.logger.WithField("path", s.statusURI)
+	current := s.Agent.Config()
+	buf, err := yaml.Marshal(current)
+	if err != nil {
+		entry.WithError(err).Warn("Cannot marshal state")
+		return err
+	}
+	writer, err := s.opener.Writer(context.Background(), s.statusURI)
```

Comment:
Could we add a timeout to each call so that we don't deadlock? ref #13816

Reply:
Thank you for the review. Done.